File: LAB6/zadanie 6.2/aes_cfb_cpp/src/aes_cfb.cpp
#include "../include/aes_cfb.h"
#include <openssl/rand.h>
#include <fstream>
#include <sstream>
#include <iomanip>

// Конструктор с автоматической генерацией ключей
AES_CFB::AES_CFB() {
    // Генерация криптографически безопасного ключа
    if (RAND_bytes(key_.data(), key_.size()) != 1) {
        throw std::runtime_error("Ошибка генерации ключа");
    }
    
    // Генерация криптографически безопасного вектора инициализации
    if (RAND_bytes(iv_.data(), iv_.size()) != 1) {
        throw std::runtime_error("Ошибка генерации вектора инициализации");
    }
}

// Шифрование данных
std::vector<uint8_t> AES_CFB::encrypt(const std::string& plaintext) {
    validate();
    const std::string padded = add_padding(plaintext);
    
    AES_KEY aes_key;
    if (AES_set_encrypt_key(key_.data(), 128, &aes_key) < 0) {
        throw std::runtime_error("Ошибка установки ключа шифрования");
    }
    
    std::vector<uint8_t> ciphertext(padded.size());
    IV ivec = iv_;
    int num = 0;
    
    AES_cfb128_encrypt(
        reinterpret_cast<const uint8_t*>(padded.data()),
        ciphertext.data(),
        padded.size(),
        &aes_key,
        ivec.data(),
        &num,
        AES_ENCRYPT
    );
    
    return ciphertext;
}

// Дешифрование данных
std::string AES_CFB::decrypt(const std::vector<uint8_t>& ciphertext) {
    validate();
    
    AES_KEY aes_key;
    if (AES_set_encrypt_key(key_.data(), 128, &aes_key) < 0) {
        throw std::runtime_error("Ошибка установки ключа дешифрования");
    }
    
    std::string plaintext(ciphertext.size(), '\0');
    IV ivec = iv_;
    int num = 0;
    
    AES_cfb128_encrypt(
        ciphertext.data(),
        reinterpret_cast<uint8_t*>(plaintext.data()),
        ciphertext.size(),
        &aes_key,
        ivec.data(),
        &num,
        AES_DECRYPT
    );
    
    return remove_padding(plaintext);
}

// Сохранение ключа в файл
void AES_CFB::save_key(const std::string& filename) const {
    std::ofstream file(filename, std::ios::binary);
    if (!file) {
        throw std::runtime_error("Ошибка открытия файла ключа для записи");
    }
    file.write(reinterpret_cast<const char*>(key_.data()), key_.size());
}

// Загрузка ключа из файла
void AES_CFB::load_key(const std::string& filename) {
    std::ifstream file(filename, std::ios::binary);
    if (!file) {
        throw std::runtime_error("Ошибка открытия файла ключа для чтения");
    }
    file.read(reinterpret_cast<char*>(key_.data()), key_.size());
}

// Преобразование байтов в HEX-строку
std::string AES_CFB::bytes_to_hex(const uint8_t* data, size_t length) {
    std::ostringstream oss;
    oss << std::hex << std::setfill('0');
    for (size_t i = 0; i < length; ++i) {
        oss << std::setw(2) << static_cast<int>(data[i]);
    }
    return oss.str();
}

// Проверка валидности ключей
void AES_CFB::validate() const {
    if (key_.size() != BLOCK_SIZE) {
        throw std::invalid_argument("Неверный размер ключа");
    }
    if (iv_.size() != BLOCK_SIZE) {
        throw std::invalid_argument("Неверный размер вектора инициализации");
    }
}

// Добавление PKCS7 заполнения
std::string AES_CFB::add_padding(const std::string& data) const {
    size_t pad_len = BLOCK_SIZE - (data.size() % BLOCK_SIZE);
    if (pad_len == 0) pad_len = BLOCK_SIZE;
    return data + std::string(pad_len, static_cast<char>(pad_len));
}

// Удаление PKCS7 заполнения
std::string AES_CFB::remove_padding(const std::string& data) const {
    if (data.empty()) return data;
    size_t pad_len = static_cast<uint8_t>(data.back());
    if (pad_len > BLOCK_SIZE || pad_len == 0) return data;
    return data.substr(0, data.size() - pad_len);
}

File: LAB6/zadanie 6.2/aes_cfb_go/aes_cfb/aes_cfb.go
/*
Пакет aescfb реализует AES-128 шифрование в режиме CFB (Cipher Feedback)

Особенности:
- Автоматическая генерация ключей
- Безопасное управление памятью
- Поддержка PKCS7 заполнения
- Полная обработка ошибок
*/
package aescfb

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/hex"
	"errors"
	"fmt"
	"io"
	"os"
)

const BlockSize = aes.BlockSize // Размер блока AES в байтах

// AESCFB представляет контекст шифрования AES-128 CFB
type AESCFB struct {
	key []byte // Ключ шифрования
	iv  []byte // Вектор инициализации
}

// New создает новый экземпляр AESCFB со случайными ключом и вектором инициализации
func New() (*AESCFB, error) {
	key, err := generateRandomBytes(BlockSize)
	if err != nil {
		return nil, fmt.Errorf("ошибка генерации ключа: %w", err)
	}

	iv, err := generateRandomBytes(BlockSize)
	if err != nil {
		return nil, fmt.Errorf("ошибка генерации вектора инициализации: %w", err)
	}

	return &AESCFB{key: key, iv: iv}, nil
}

// Encrypt шифрует данные с использованием AES-128 CFB
func (a *AESCFB) Encrypt(plaintext []byte) ([]byte, error) {
	if err := a.validate(); err != nil {
		return nil, err
	}

	block, err := aes.NewCipher(a.key)
	if err != nil {
		return nil, fmt.Errorf("ошибка создания шифра: %w", err)
	}

	padded := addPadding(plaintext, BlockSize)
	ciphertext := make([]byte, len(padded))
	stream := cipher.NewCFBEncrypter(block, a.iv)
	stream.XORKeyStream(ciphertext, padded)

	return ciphertext, nil
}

// Decrypt дешифрует данные с использованием AES-128 CFB
func (a *AESCFB) Decrypt(ciphertext []byte) ([]byte, error) {
	if err := a.validate(); err != nil {
		return nil, err
	}

	block, err := aes.NewCipher(a.key)
	if err != nil {
		return nil, fmt.Errorf("ошибка создания шифра: %w", err)
	}

	plaintext := make([]byte, len(ciphertext))
	stream := cipher.NewCFBDecrypter(block, a.iv)
	stream.XORKeyStream(plaintext, ciphertext)

	return removePadding(plaintext), nil
}

// Key возвращает текущий ключ шифрования
func (a *AESCFB) Key() []byte {
	return a.key
}

// IV возвращает текущий вектор инициализации
func (a *AESCFB) IV() []byte {
	return a.iv
}

// SaveKey сохраняет ключ в файл
func (a *AESCFB) SaveKey(filename string) error {
	return os.WriteFile(filename, a.key, 0600)
}

// LoadKey загружает ключ из файла
func (a *AESCFB) LoadKey(filename string) error {
	key, err := os.ReadFile(filename)
	if err != nil {
		return fmt.Errorf("ошибка чтения файла ключа: %w", err)
	}
	
	if len(key) != BlockSize {
		return errors.New("неверный размер ключа")
	}
	
	a.key = key
	return nil
}

// BytesToHex преобразует байты в шестнадцатеричную строку
func BytesToHex(data []byte) string {
	return hex.EncodeToString(data)
}

// validate проверяет корректность ключа и вектора инициализации
func (a *AESCFB) validate() error {
	if len(a.key) != BlockSize {
		return errors.New("неверный размер ключа")
	}
	if len(a.iv) != BlockSize {
		return errors.New("неверный размер вектора инициализации")
	}
	return nil
}

// generateRandomBytes генерирует криптографически безопасные случайные байты
func generateRandomBytes(n int) ([]byte, error) {
	b := make([]byte, n)
	if _, err := io.ReadFull(rand.Reader, b); err != nil {
		return nil, err
	}
	return b, nil
}

// addPadding добавляет PKCS7 заполнение к данным
func addPadding(data []byte, blockSize int) []byte {
	padding := blockSize - (len(data) % blockSize)
	if padding == 0 {
		padding = blockSize
	}
	padded := make([]byte, len(data)+padding)
	copy(padded, data)
	for i := len(data); i < len(padded); i++ {
		padded[i] = byte(padding)
	}
	return padded
}

// removePadding удаляет PKCS7 заполнение из данных
func removePadding(data []byte) []byte {
	if len(data) == 0 {
		return data
	}
	padding := int(data[len(data)-1])
	if padding > len(data) || padding == 0 {
		return data
	}
	return data[:len(data)-padding]
}
